File: dp/121.go
func maxProfit(prices []int) int {
    
    if len(prices) == 0 {
        return 0
    }
    ret := 0
    curlow := prices[0]
    for i := 1; i < len(prices); i++ {
        if prices[i] - curlow > ret {
            ret = prices[i] - curlow
        }
        if prices[i] < curlow {
            curlow = prices[i]
        }
    }
    return ret
}


func maxProfit(prices []int) int {
    if len(prices) == 0 {
        return 0
    }
    dp := make([][]int, len(prices))
    for i, _ := range dp {
        dp[i] = make([]int, 2)
    }
    // 0表示未持有 1表示持有
    dp[0][0] = 0
    dp[0][1] = -prices[0]
    for i := 1; i < len(prices); i++ {
        dp[i][0] = max(dp[i-1][0], dp[i-1][1]+prices[i])
        dp[i][1] = max(dp[i-1][1], -prices[i])
    }
    return dp[len(prices)-1][0]
}

func max(a int, b int) int {
    if a < b {
        return b
    }else {
        return a
    }
}

File: dp/121.cc
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.empty()) {
            return 0;
        }
        int ret = 0;
        int low = prices[0];
        for (int i = 1; i < prices.size(); i++) {
            if (low > prices[i]) {
                low = prices[i];
            }
            if (ret < (prices[i]-low)) {
                ret = prices[i] - low;
            }
        }
        return ret;
    }
};


class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.empty())
            return 0;
        int hold0 = 0;
        int nothold0 = 0;
        int hold1 = -prices[0];
        int nothold1 = 0;
        for (size_t i = 1; i < prices.size(); ++i) {
            nothold0 = std::max(hold1+prices[i], nothold1);
            // hold0 = std::max(nothold1-prices[i], hold1); 这种 case 是可以买多次的
            hold0 = std::max(-prices[i], hold1);
            hold1 = hold0;
            nothold1 = nothold0;
        }
        return nothold0;
    }
};

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int cur = INT_MAX, ret = 0;
        for (const auto& price : prices) {
            ret = std::max(ret, price - cur);
            cur = std::min(cur, price);
        }
        return ret;
    }
};
